### ISSUE TITLE
IReadOnlyList<T> for options list

### ISSUE THREAD

**Issue by MgSam** [commandlineparser/commandline#668]
I'd like to suggest list options support more than just `IEnumerable<T>` properties. `IEnumerable<T>` is not a great choice for list properties because it hides information about the underlying collection (order, count). 

Adding support for `IReadOnlyList<T>` would be very helpful because it is both ordered (supports an indexer), and has a count. Support for arrays, `ICollection<T>`, `List<T>`, `IList<T>`, `IReadOnlyCollection<T>` are also things that could be considered.

**Comment by moh-hassan** [comment]
Thanks @MgSam for the suggestion.

`IEnumerable<T>`  values may be passed as a parameter to other method for processing.
The [Guidelines for Collections](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/guidelines-for-collections) advice: 
>DO use the least-specialized type possible as a parameter type. Most members taking collections as parameters use the IEnumerable<T> interface.

Also, `IEnumerable<T>` suport net40+ and   `IReadOnlyList<T>`  support net45+
`IEnumerable<T>` can be converted to other collections  as needed.

You can, if you wish, to use `IReadOnlyList<T>` and it's working  because it [inherit ](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ireadonlylist-1?view=netcore-3.1)from `IEnumerable<T>`

**Comment by blackboxlogic** [comment]
`IReadOnlyList<T>` and `IList<T>` work, You said "because it inherits from `IEnumerable<T>`".
Why don't options of type `List<T>` work?

**Comment by teneko** [comment]
Yep, if you follow the quoted guidance further I would like the idea of a list factory, a factory of those lists that are used in options, so the use of such concrete list or one of his declared base classes/interfaces like IEnumerable<T> in the options is comprehensible. Furthermore it must have documented which interfaces are supported, because we, the consumer, do not know of which type the concrete list is.

It annoys me alot to think about, whether the IEnumerator<T> from IEnumerable<T>.GetEnumerator is resettable or not: it forces me literally to do `new List<T>(oneOfTheseSequence)`, because its concrete type can not be even a concrete type but a (yield) iterator method. In my opinion this is a bad design decision, because as INPUT (parameters) it's completely fine to take the smallest interface but as OUTPUT (fields, especially public)? Think about it:

```
interface IEnumerable<T> {}
class List<T> : IEnumerabe<T> {}
IEnumerable<T> ParseSequence<T>() => new List<T>() {{default(T)}};
```

Would you really just give the enumerable the consumer of your API? xD

If your mutable list is member of your business logic (I mean used again after it has been created AND was accessible from consumer) then it's completely fine to expose it by IReadOnlyList or wrap it preferred by ReadOnlyCollection<T> (no copy, only wrap) and expose this one instead, when not, then expose the concrete list, why care about it what the consumer is doing with it?
